app/controllers/orders_controller.rb

```diff
@@ -48,28 +80,89 @@ def new
   def update
     @order = Order.find(session[:order_id]) #session is persistent, the cookie has the session information.
     @order.update(order_param[:order]) # when you get a request, here is my information for right now.
+
     reduce_inventory(@order)
-    if @order.save
+
+    zip = params["order"]["zip"]
+    city = params["order"]["city"]
+    state = params["order"]["state"]
+    # message = ""
+    message = valid_location(zip, city, state)
+
+
+    if @order.save && message == ""
       redirect_to order_path
     else
+      flash.now[:message] = message if !message.empty?
       render :edit
     end
   end
 
+
+  def valid_location(zip, city, state)
+    zips = valid_zip(zip)
+    message = ""
+
+    if !zips.nil?
+      zip_city = zip.to_region(:city => true)
+      zip_state = zip.to_region(:state => true)
+
+      valid_city = city.downcase.strip != zip_city.downcase
+      valid_state = state.upcase != zip_state
+      message+=" **City does not match zip code. Did you mean: #{zip_city}? " if valid_city
+      message += " **State does not match zip code and or 2 letter abbreviation. Did you mean: #{zip_state}? "if valid_state
+      # flash[:success] = message
+    elsif zips.nil?
+      message += "Invalid Zip Code!"
+      # flash[:success] = message
+    end
+    # raise
```

Comment:
This was a slick feature - well done.
Make sure you clean up your commented out code.